django/thunderstore/api/cyberstorm/views/packages.py

```diff
@@ -0,0 +1,467 @@
+
+        # PageNotAnInteger error won't be raised here since deserializer
+        # has already checked that the page parameter is an integer.
+        try:
+            page = paginator.page(params["page"])
+        except EmptyPage:
+            raise ParseError("Page index error: no results on requested page")
+
+        return page
+
+    def _serialize_results(
+        self,
+        params: OrderedDict,
+        package_page: Page,
+    ) -> PackageListResponseSerializer:
+        """
+        Format results to transportation.
+        """
+        packages = [
+            {
+                "categories": p.community_listings.all()[0].categories.all(),
+                "community_identifier": p.community_listings.all()[
+                    0
+                ].community.identifier,
+                "description": p.latest.description,
+                "download_count": p.download_count,
+                "icon_url": p.latest.icon.url if bool(p.latest.icon) else None,
+                "is_deprecated": p.is_deprecated,
+                "is_nsfw": p.community_listings.all()[0].has_nsfw_content,
+                "is_pinned": p.is_pinned,
+                "last_updated": p.date_updated,
+                "namespace": p.namespace.name,
+                "name": p.name,
+                "rating_count": p.rating_count,
+                "size": p.latest.file_size,
+            }
+            for p in package_page.object_list
+        ]
+
+        (prev_url, next_url) = self._get_sibling_pages(params, package_page)
+
+        return PackageListResponseSerializer(
+            {
+                "count": package_page.paginator.count,
+                "previous": prev_url,
+                "next": next_url,
+                "results": packages,
+            },
+        )
+
+    def _get_full_cache_vary(self, params: OrderedDict) -> str:
+        """
+        Return cache key for CachedPaginator.
+        """
+        cache_vary = self._get_paginator_cache_vary_prefix()
+        cache_vary += f".{params['deprecated']}"
+        cache_vary += f".{params['nsfw']}"
+        cache_vary += f".{sorted(params['included_categories'])}"
+        cache_vary += f".{sorted(params['excluded_categories'])}"
+        cache_vary += f".{params.get('section', '-')}"
+        cache_vary += f".{params.get('q', '-')}"
+        cache_vary += f".{params['ordering']}"
+        cache_vary += f".{params['page']}"
+        cache_vary += f".{self.page_size}"
+        return cache_vary
+
+    def _get_sibling_pages(
+        self,
+        params: OrderedDict,
+        package_page: Page,
+    ) -> Tuple[Optional[str], Optional[str]]:
+        """
+        Return the URLs to previous and next pages of this result set.
+        """
+        for value in params.values():
+            if type(value) == list:
+                value.sort()
+
+        request_url = self.request.get_raw_uri()
+        base_url = request_url.split("?")[0] if "?" in request_url else request_url
```

Comment:
Ideally should build the full URL without relying on the request URL as that's a potential attack vector which can bleed over to other users via cache. There's a few examples in the codebase for full URL building without relying on the request.